src/group_theory/free_product.lean

```diff
@@ -0,0 +1,150 @@
+`free_product.rel` is confluent.
+
+## References
+
+[van der Waerden, *Free products of groups*][MR25465]
+
+-/
+
+variables {ι : Type*} (M : Π i : ι, Type*) [Π i, monoid (M i)]
+
+/-- A relation on the free monoid on alphabet `Σ i, M i`, relating `⟨i, 1⟩` with `1` and
+`⟨i, x⟩ * ⟨i, y⟩` with `⟨i, x * y⟩`. -/
+inductive free_product.rel : free_monoid (Σ i, M i) → free_monoid (Σ i, M i) → Prop
+| of_one (i : ι) : free_product.rel (free_monoid.of ⟨i, 1⟩) 1
+| of_mul {i : ι} (x y : M i) : free_product.rel (free_monoid.of ⟨i, x⟩ * free_monoid.of ⟨i, y⟩)
+  (free_monoid.of ⟨i, x * y⟩)
+
+/-- The free product (categorical coproduct) of an indexed family of monoids. -/
+@[derive [monoid, inhabited]]
+def free_product : Type* := (con_gen (free_product.rel M)).quotient
+
+namespace free_product
+
+variable {M}
+
+/-- The inclusion of a summand into the free product. -/
+def of {i : ι} : M i →* free_product M :=
+{ to_fun   := λ x, con.mk' _ (free_monoid.of $ sigma.mk i x),
+  map_one' := (con.eq _).mpr (con_gen.rel.of _ _ (free_product.rel.of_one i)),
+  map_mul' := λ x y, eq.symm $ (con.eq _).mpr (con_gen.rel.of _ _ (free_product.rel.of_mul x y)) }
+
+lemma of_apply {i} (m : M i) : of m = con.mk' _ (free_monoid.of $ sigma.mk i m) := rfl
+
+variables {N : Type*} [monoid N]
+
+/-- See note [partially-applied ext lemmas]. -/
+@[ext] lemma ext_hom (f g : free_product M →* N) (h : ∀ i, f.comp (of : M i →* _) = g.comp of) :
+  f = g :=
+(monoid_hom.cancel_right con.mk'_surjective).mp $ free_monoid.hom_eq $ λ ⟨i, x⟩,
+  by rw [monoid_hom.comp_apply, monoid_hom.comp_apply, ←of_apply,
+    ←monoid_hom.comp_apply, ←monoid_hom.comp_apply, h]
+
+/-- A map out of the free product corresponds to a family of maps out of the summands. This is the
+universal property of the free product, charaterizing it as a categorical coproduct. -/
+@[simps symm_apply]
+def lift : (Π i, M i →* N) ≃ (free_product M →* N) :=
+{ to_fun := λ fi, con.lift _ (free_monoid.lift $ λ p : Σ i, M i, fi p.fst p.snd) $ con.con_gen_le
+    begin
+      simp_rw [con.rel_eq_coe, con.ker_rel],
+      rintros _ _ (i | ⟨i, x, y⟩),
+      { change free_monoid.lift _ (free_monoid.of _) = free_monoid.lift _ 1,
+        simp only [monoid_hom.map_one, free_monoid.lift_eval_of], },
+      { change free_monoid.lift _ (free_monoid.of _ * free_monoid.of _) =
+          free_monoid.lift _ (free_monoid.of _),
+        simp only [monoid_hom.map_mul, free_monoid.lift_eval_of], }
+    end,
+  inv_fun := λ f i, f.comp of,
+  left_inv := by { intro fi, ext i x,
+    rw [monoid_hom.comp_apply, of_apply, con.lift_mk', free_monoid.lift_eval_of], },
+  right_inv := by { intro f, ext i x,
+    simp only [monoid_hom.comp_apply, of_apply, con.lift_mk', free_monoid.lift_eval_of], } }
+
+@[simp] lemma lift_of {N} [monoid N] (fi : Π i, M i →* N) {i} (m : M i) :
+  lift fi (of m) = fi i m :=
+by conv_rhs { rw [←lift.symm_apply_apply fi, lift_symm_apply, monoid_hom.comp_apply] }
+
+lemma induction_on {C : free_product M → Prop}
```

Comment:
This maybe wants
```suggestion
@[elab_as_eliminator]
lemma induction_on {C : free_product M → Prop}
```
I don't remember which of the `induction_on` and `induction` lemmas should have that.